### PR TITLE
Some miscellaneous MinimalLib enhancements

### PR DESCRIPTION
**Pull request by ptosco**
This PR contributes back some MinimalLib enhancements that I have been using for some time, namely:
- expose `SubstructureLibrary` to MinimalLib
- expose `acceptFailure` flag to MinimalLib `generate_aligned_coords`
- expose `merge_query_hs` functionality to MinimalLib
- expose `kekulize` flag to MinimalLib `get_mol`
- add relevant JS tests
Exposing `SubstructLibrary` enables substructure search functionalty in the browser. In this context, the `merge_query_hs` functionality proves useful to incorporate explicit Hs as queries to forbid substituents as certain positions.
Exposing the `acceptFailure` flag avoids that existing coordinates get overwritten when the attempt to generate constrained 2D coordinates fails.
Finally, exposing the `kekulize` flag allows to depict things like R-groups with partial aromatic rings avoiding kekulization failures without need to call `get_qmol` instead of `get_mol`.

### CODE REVIEW

**Comment by greglandrum** [comment]
@ptosco : there's a lot of stuff in here. It may take me a few days

**Comment by ptosco** [comment]
@greglandrum No worries. Note that the big change in `common.h` is mostly due to CRLF->LF conversion; the only real change is the addition of a `details_json.kekulize` option. to opt in/out kekulization from `SanitizeOps`.

**Comment by greglandrum** [comment]
@ptosco I suspect that the conflict is due to the fact that you also changed the line endings in that other PR.
If you can fix the conflict I'll go ahead and finish reviewing

File: Code/MinimalLib/tests/tests.js [hunk 1]
```diff
@@ -214,16 +333,109 @@ M  END`;
...
+function test_get_mol_no_kekulize() {
+    var mol = RDKitModule.get_mol("c");
+    assert(!mol.is_valid());
+    mol = RDKitModule.get_mol("c", JSON.stringify({kekulize: false}));
+    assert(mol.is_valid());
+}
+
+
 initRDKitModule().then(function(instance) {
+    var done = {
```
Comment (greglandrum) [inline Code/MinimalLib/tests/tests.js:407]: apologies for my JS-illiteracy: why is this here?

**Review by greglandrum — CHANGES_REQUESTED** [review]
Some questions and comments are inline here.
I have a couple of other requests that I'll make as separate threads.

**Comment by greglandrum** [comment]
Inspired by, but not part of this PR: we should also expose AdjustQueryParameters() in the minimallib: #4187

**Comment by greglandrum** [comment]
@ptosco : a suggestion, please let me know what you think
I think we should require that documentation be added to the GettingStartedInJS document when we add new functionality to the JS wrapper. I'd like to avoid having this thing be largely undocumented and it seems like the only way to ensure that things stay up-to-date is to make it part of the process when adding new features.
What do you think?

**Comment by greglandrum** [comment]
It doesn't need to be added to this PR, but we should also support merge_query_hs in the CFFI wrapper.

File: Code/MinimalLib/minilib.cpp [hunk 2]
```diff
@@ -310,19 +318,94 @@ std::string JSMol::generate_aligned_coords(const JSMol &templateMol,
...
+              new CachedTrustedSmilesMolHolder()),
+          boost::shared_ptr<PatternHolder>(new PatternHolder()))),
+      d_num_bits(num_bits) {
+  d_molHolder = dynamic_cast<CachedTrustedSmilesMolHolder *>(
+      d_sslib->getMolHolder().get());
+  d_fpHolder = dynamic_cast<PatternHolder *>(d_sslib->getFpHolder().get());
+}
+
+int JSSubstructLibrary::add_trusted_smiles(const std::string &smi) {
+  std::unique_ptr<RWMol> mol(SmilesToMol(smi));
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/MinimalLib/minilib.cpp]: Oh, that was indeed a major oversight, thanks for pointing it out. Fixed now. The C++ code in `SubstructLibrary` calls `ROMol::updatePropertyCache(true)`, so I did the same for consistency. `PatternFingerprintMol` takes care of calling `MolOps::fastFindRings`, so no need to call it in `JSSubstructLibrary::add_trusted_smiles`.

File: Code/MinimalLib/minilib.cpp [hunk 3]
```diff
@@ -17,12 +17,14 @@
 #include <GraphMol/RDKitBase.h>
 #include <GraphMol/MolPickler.h>
 #include <GraphMol/SmilesParse/SmilesParse.h>
-#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/SmilesParse/SmartsWrite.h>
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/MinimalLib/minilib.cpp]: Just leftovers from a previous version that had some debugging output. Reverted.

File: Code/MinimalLib/tests/tests.js [hunk 4]
```diff
@@ -10,12 +10,18 @@
 //
 
 const assert = require('assert');
+const {
+    performance
+  } = require('perf_hooks');
 var initRDKitModule = require("../demo/RDKit_minimal.js");
 var RDKitModule;
+const fs       = require('fs');
+const zlib     = require('zlib');
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/MinimalLib/tests/tests.js]: `zlib` is a leftover from a previous incarnation of the test where I was reading a large `gzip`-compressed SMILES file. Removed. `fs` is used to parse a SMILES file.

File: Code/MinimalLib/tests/tests.js [hunk 5]
```diff
@@ -133,8 +168,91 @@ function test_abbreviations(){
     assert.equal(bmol.get_cxsmiles(),"*C1CCC1 |$CF3;;;;$|");
 }
 
+function test_substruct_library(done){
+    var smiReader = readline.createInterface({
+        input: fs.createReadStream(__dirname + '/../../GraphMol/test_data/compounds.smi')
+    });
+    var sslib = new RDKitModule.SubstructLibrary();
+    var t0 = performance.now()
+    console.log('Started adding trusted SMILES');
```
[the remark this answers is not in the document]
Comment (ptosco) [reply Code/MinimalLib/tests/tests.js]: Agreed, commented out.

File: Code/MinimalLib/tests/tests.js (hunk 1 above)
Comment (ptosco) [reply Code/MinimalLib/tests/tests.js]: Reading SMILES from file is asynchronous, so I need to poll for completion to avoid printing "Test completed" before the test is actually completed.

**Review by greglandrum — APPROVED** [review]
LGTM
